src/webgpu/api/validation/vertex_access.spec.ts

```diff
@@ -1,91 +1,483 @@
->         - An unused {index (with uselessly small range), vertex} buffer
->           is bound (immediately before draw call)
->         - }
->     - All indexed (in/direct) draw commands, {
->         - No index buffer is bound
->         - Same GPUBuffer bound to index buffer and a vertex buffer slot
->             - Non-overlapping, overlapping ranges
->         - Draw call needs to read {=, >} the bound index buffer range
->           (with GPUBuffer that is always large enough)
->             - range is too small and GPUBuffer is large enough
->             - range and GPUBuffer are exact size
->             - x= all index formats
->         - Bound index buffer range is significantly larger than necessary
->         - }
->     - Alignment constraints on setVertexBuffer if any
->     - Alignment constraints on setIndexBuffer if any
-> - In queue.submit():
->     - Indexed draw call with index buffer containing:
->         - Index value that goes out-of-bounds on a bound vertex buffer range
->         - Index value that is extremely large (but not the primitive restart value)
->     - Line strip or triangle strip with index buffer containing:
->         - Primitive restart value
->         - Primitive restart value minus one (and the bound vertex buffers are < that size)
->     - Indirect draw call with arguments that:
->         - Go out-of-bounds on the bound index buffer range
->         - Go out-of-bounds on the bound vertex buffer range
-
-TODO: Had two plans with roughly the same name. Figure out where to categorize these notes:
-> All x= {render pass, render bundle}
->
-> - non-indexed draws:
->     - vertex access out of bounds (make sure this doesn't overlap with robust access)
->         - bound vertex buffer **ranges** are {exact size, just under exact size} needed for draws with:
->             - vertexCount largeish
->             - firstVertex {=, >} 0
->             - instanceCount largeish
->             - firstInstance {=, >} 0
->         - include VBs with both step modes
->     - x= {draw, drawIndirect}
-> - indexed draws:
->     - vertex access out of bounds (make sure this doesn't overlap with robust access)
->         - bound vertex buffer **ranges** are {exact size, just under exact size} needed for draws with:
->             - a vertex index in the buffer is largeish
->             - baseVertex {=, >} 0
->             - instanceCount largeish
->             - firstInstance {=, >} 0
->         - include VBs with both step modes
->     - x= {drawIndirect, drawIndexedIndirect}
+Here we test the vertex buffer validation within creating render pipeline, setting index and vertex
+buffer, and calling draw function. For draw function, we only tests OOB for draw, and index buffer
+and instance step mode vertex buffer OOB for drawIndexed. Vertex step mode vertex buffer OOB for
+drawIndexed, vertex buffer OOB for drawIndirect and vertex and index buffer OOB for
+drawIndexedIndirect are covered in robust access.
```

Comment:
Sorry I should have started by reviewing the test and test plan instead of the test support code as I realized there are some more fundamental issues here. Apologies to have assigned you such a messy task, I forgot how messy it was!

A lot of these validation rules (like usages and validity) are checked in api/validation/encoding/cmds/render/setIndexBuffer.spec.ts and setVertexBuffer.spec.ts. I think some are also covered in vertex_state.spec.ts. Vertex buffer OOB is covered in robust_access_vertex.spec.ts.

This is one reason why these notes needed a review: a lot of them are actually covered elsewhere now. These notes also speculate about a lot of possible validation rules that turned out not to exist, so there should be no validation tests for them (e.g. overlap between a vertex buffer and an index buffer binding). The author (pretty sure it was me) also often got confused about what should be validation tests vs. what should be operation tests.

These are the major validation tests from this plan that I think we were still missing - at least, I can't find them:
- `draw[Indexed][Indirect]()` without vertex buffer
- `drawIndexed[Indirect]()` without index buffer
- `draw()` out of bounds of vertex buffer binding range
- `drawIndexed()` out of bounds of index buffer binding range

The following also seem to be missing, but are operation tests as they do not generate validation errors. These need to be added as TODOs to their appropriate files or directories so they're tracked.
- `drawIndirect()` out of bounds of vertex buffer binding range
- `drawIndexedIndirect()` out of bounds of index buffer binding range

See below for more notes on where tests should be categorized. I have many notes that mention I think things are already tested - but I need you to double check that before deleting new duplicates.

Ultimately as tests get moved to other files, this file may not exist - it has a very vague name and all of these tests can probably be better categorized.

Reply:
I am moving missing buffer test to `all_needed_buffer_should_be_bound` in `api/validation/encoding/cmds/render/state_tracking.spec.ts`, tests relating with index or vertex buffer OOB when draw[Indexed] into `api/validation/encoding/cmds/render/direct_draw.spec.ts` (new created),  vertex buffer layouts validation test for creating render pipeline into `api/validation/createRenderPipeline.spec.ts`, and largeish buffer test into `api/operation/rendering/draw.spec.ts`.
The original `vertex_access.spec.ts` is now empty and removed.